### PR TITLE
fix(channels): clear stale OAuth Connecting badges across auth modes (#2128)

### LINKED ISSUE #2128: OAuth connection badges get stuck in Connecting
## Summary

OAuth connection badges can get stuck in `Connecting`, and starting a different OAuth method does not clear the prior pending state.

## Problem

In the Discord connection panel, both `OAuth Sign-in` and `Login with OpenHuman` can appear pinned to `Connecting` at the same time. A similar issue was also observed across GitHub and GitLab: when a new OAuth flow starts, the previous method’s pending state is not cleared, so multiple methods look like they are mid-connect simultaneously.

Expected behavior: each OAuth method should have accurate connection status. Starting a new OAuth flow should either cancel/clear the previous pending state or scope pending state to the specific method/account so stale `Connecting` badges do not remain visible.

Impact: users cannot tell which OAuth flow is active, whether a previous attempt failed, or which connection method needs attention. This makes integration setup feel stuck and unreliable.

Steps to reproduce:
1. Open OpenHuman and go to a connection panel with multiple OAuth methods, such as Discord.
2. Start one OAuth method, such as `OAuth Sign-in`.
3. Before it fully resolves, start another method, such as `Login with OpenHuman`, or return after an OAuth failure.
4. Observe both methods can remain shown as `Connecting`.
5. Repeat on GitHub/GitLab connection flows and observe similar stale pending status.

Version / platform: OpenHuman app connection/OAuth UI, observed May 18, 2026 from user screenshot and report.

## Solution (optional)

Investigate the connection status state model, likely in `channelConnections` or the equivalent per-method status tracking. Ensure pending OAuth state is keyed by provider + method + attempt ID, and that starting, completing, failing, or cancelling a flow clears stale pending status for prior attempts.

## Acceptance criteria

- [ ] **Repro gone** — Multiple OAuth methods no longer remain stuck as `Connecting` after starting a new OAuth flow or after a prior failure.
- [ ] **Pending state scoped** — OAuth pending state is scoped to the active provider/method/attempt and does not leak across other methods.
- [ ] **Prior state cleared** — Starting a new OAuth method clears or cancels the previous method’s pending state where appropriate.
- [ ] **Failure state shown** — Failed OAuth attempts transition to a clear failure/retry state instead of staying in `Connecting`.
- [ ] **Provider coverage** — The fix covers Discord, GitHub, GitLab, and other multi-method OAuth connection panels using the same state path.
- [ ] **Diagnostics captured** — Logs include safe provider, method, attempt ID, status transitions, and failure reason without exposing tokens or OAuth secrets.
- [ ] **Regression safety** — Unit, integration, or E2E coverage verifies overlapping OAuth attempts, failure cleanup, and status isolation across methods/providers.
- [ ] **Diff coverage ≥ 80%** — the fix PR meets the changed-lines coverage gate (Vitest + cargo-llvm-cov, enforced by [`.github/workflows/coverage.yml`](../../.github/workflows/coverage.yml)).

## Related

- User-provided screenshot from May 18, 2026 showing connection failure UI and related stuck connection state report.

### LINKED ISSUE #2141: Channel selector hides channel error state and falls back to Disconnected
Channel selector hides channel errors as disconnected

## Summary

When a Telegram or Discord channel connection fails, the channel detail panel can show an error state, but the channel selector tab still shows the channel as `Disconnected`.

This hides real setup failures in the overview and makes the channel look inactive instead of broken.

## Problem

The channel selector currently aggregates channel status by checking for `connected`, then `connecting`, and otherwise falling back to `disconnected`. It does not surface `error`, even though the Telegram and Discord channel config panels explicitly write `error` states for validation failures, timeouts, and other connection problems.

Expected behavior: if a channel has no connected or connecting auth mode, but at least one auth mode is in `error`, the selector should show `Error`.

Actual behavior: the selector shows `Disconnected`, which makes failed channel setups look like nothing is wrong.

Impact: users can miss active connection failures in the top-level Channels overview and may not realize they need to return to a specific channel panel to retry or inspect the error.

Steps to reproduce:
1. Open the Channels page in the desktop app.
2. Go to Telegram or Discord.
3. Trigger a connection failure, for example by submitting invalid required fields or waiting for a managed-link flow to time out.
4. Confirm that the channel detail panel shows an error state or error message.
5. Look at the channel selector tabs at the top.
6. Observe that the affected channel is shown as `Disconnected` instead of `Error`.

Version / platform: current local `main` checkout on macOS desktop app, observed from the React channel UI code path in `app/src/components/channels/*`.

## Solution (optional)

The likely fix is in ChannelSelector.tsx. Its channel-level status aggregation should account for `error` before falling back to `disconnected`.

A simple expected priority would be:
- `connected`
- `connecting`
- `error`
- `disconnected`

It would also be worth adding a selector test that seeds channel state with an auth mode in `error` and verifies that the tab renders `Error`.

## Acceptance criteria

- [ ] **Repro gone** — A failed Telegram or Discord connection no longer appears as `Disconnected` in the channel selector when the underlying auth mode is in `error`.
- [ ] **Correct priority** — Channel-level status aggregation prefers `connected`, then `connecting`, then `error`, and only falls back to `disconnected` when no other state is present.
- [ ] **Regression safety** — Unit coverage is added or updated for `ChannelSelector` so channel error aggregation is tested.
- [ ] **Diff coverage ≥ 80%** — the fix PR meets the changed-lines coverage gate (Vitest + cargo-llvm-cov, enforced by `.github/workflows/coverage.yml`).
- [ ] **Cross-channel check** — The behavior is verified for both Telegram and Discord error states, since both panels already write `error` into channel connection state.

## Related

- Related but not identical: #2128, which covers OAuth connection badges getting stuck in `Connecting`
- Relevant code:
  - ChannelSelector.tsx
  - TelegramConfig.tsx
  - DiscordConfig.tsx
- Current selector tests:
  - ChannelSelector.test.tsx

### PR DESCRIPTION
**Pull request by sanil-23**
## Summary

- Centralises OAuth deep-link → channel-badge transitions behind a new
  `useOAuthConnectionListener` hook so every channel panel handles both
  `oauth:success` and `oauth:error` consistently.
- Adds a `clearOtherPendingForChannel` reducer so starting a connect flow on
  one auth mode drops any sibling auth mode that's still mid-`connecting` on
  the same channel.
- Wires `DiscordConfig` and `TelegramConfig` onto the shared hook; future
  channels with an OAuth auth mode inherit correct pending-state transitions
  automatically.
- Covers the new reducer (4 cases) and hook (8 cases) with Vitest.

## Problem

OAuth badges on the channel connection panels could get pinned at
`Connecting` indefinitely (issue #2128):

- `DiscordConfig` had a per-component `oauth:success` listener but no
  `oauth:error` listener — failed OAuth attempts never transitioned the badge
  out of `connecting`.
- `TelegramConfig` had neither — completed *and* failed OAuth attempts left
  the badge pinned.
- Both panels set `connecting` on the chosen auth mode but never cancelled
  any sibling auth mode that was already pending. Triggering a second OAuth
  method on Discord (`OAuth Sign-in` then `Login with OpenHuman`, or the
  reverse) left both methods badged `Connecting` simultaneously.

This is the exact repro from the issue. The same shape was visible across
GitHub/GitLab style multi-method panels because the underlying state model
(`channelConnections`, keyed by `(channel, authMode)`) had no notion of
mutual exclusion.

## Solution

**Shared listener hook** —
[`app/src/hooks/useOAuthConnectionListener.ts`](app/src/hooks/useOAuthConnectionListener.ts)
subscribes to both `oauth:success` and `oauth:error` window events
(dispatched from `utils/desktopDeepLinkListener.ts`), filters by `toolkit` /
`provider` case-insensitively, and dispatches the matching slice action.
Per-channel panels mount it once with `{ channel, authMode }`; cleanup on
unmount is deterministic. New channels with an OAuth auth mode inherit the
behaviour without copying any logic.

**Pending-state cancellation reducer** — `clearOtherPendingForChannel({ channel,
exceptAuthMode })` in `channelConnectionsSlice.ts` walks the auth-mode map
for one channel and transitions every `connecting` row (except the
exception) to `disconnected` with `lastError: undefined`. Cancelled rows go
to `disconnected` rather than `error` so the UI doesn't surface a misleading
failure — the user explicitly switched methods, they didn't experience an
error.

**Per-panel wiring** — `DiscordConfig` and `TelegramConfig` each:

1. Mount `useOAuthConnectionListener({ channel: <name>, authMode: 'oauth' })`
   at the top of the component (replacing the bespoke effect on Discord;
   net-new on Telegram).
2. Dispatch `clearOtherPendingForChannel` at the start of `handleConnect`
   *before* setting their own auth mode to `connecting`.

**Tradeoffs**

- The cancellation transition is `disconnected`, not a new `cancelled` state.
  Adding a dedicated state would expand the `ChannelConnectionStatus` union
  across many call sites for marginal UX value.
- The deep-link CustomEvent payload (`{ integrationId, toolkit }` for
  success, `{ provider, errorCode, message }` for error) is unchanged, so
  no symmetric change in the Tauri-side handler is needed.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) per [Testing Strategy](../gitbooks/developing/testing-strategy.md#failure-path-requirement) — 12 new Vitest cases (4 reducer + 8 hook) covering success, error, mismatched channel, mismatched provider, missing error message, custom capabilities, unsubscribe on unmount, and three sibling-cancellation shapes.
- [x] **Diff coverage ≥ 80%** — frontend-only change; `pnpm test:coverage` locally over the new files reaches 100% on changed lines (every branch in the hook + reducer is exercised by the suite).
- [x] Coverage matrix updated — `N/A: behaviour-only fix on existing surfaces (channel connection pending state)`.
- [x] All affected feature IDs from the matrix are listed in the PR description under `## Related` — `N/A: no feature ID changes`.
- [x] No new external network dependencies introduced — purely in-app state plumbing.
- [x] Manual smoke checklist updated if this touches release-cut surfaces — `N/A: no release-cut surface touched (channels panel is part of the always-shipped settings UX)`.
- [x] Linked issue closed via `Closes #NNN` in the `## Related` section — see below.

## Impact

- **Desktop only** — no mobile/web/CLI impact. The deep-link event source
  (`desktopDeepLinkListener.ts`) is Tauri-gated; the hook is a no-op outside
  Tauri because no deep-link events fire.
- **No persistence shape change** — `channelConnections` slice schema
  (`SCHEMA_VERSION = 1`) is unchanged. The new reducer only mutates existing
  rows; no migration needed.
- **No security implications** — the listener filters strictly by channel
  identifier and never reads tokens. Existing `[DeepLink][oauth:*]` logs
  remain the canonical diagnostic surface; the hook adds its own
  `channels:oauth-listener` debug namespace per the project's
  verbose-diagnostics rule.

## Related

- Closes: #2128
- Follow-up PR(s)/TODOs: none

## Provider coverage

The issue body mentions Discord, GitHub, and GitLab. The Channels page in this codebase only exposes three multi-method channel-config panels today: `DiscordConfig.tsx`, `TelegramConfig.tsx`, and `WebChannelConfig.tsx` (the last is not OAuth-driven). There is no `GitHubConfig.tsx` / `GitLabConfig.tsx` — verified via `find app/src -name "*Config.tsx"`.

GitHub OAuth does appear elsewhere in the app, but on different state slices that this PR's `channelConnections`-bound hook does not (and should not) touch:

| Surface | File(s) | State path | This PR applies? |
|---|---|---|---|
| App-level sign-in | `BootCheckGate.tsx`, OAuth callback | `deepLinkAuth` slice | No — different slice. App-level OAuth's hot-instance issue is the family fixed by #2228 / #2229. |
| Skill OAuth install | `InstallSkillDialog.tsx`, `services/api/skillsApi.ts` | skills-domain state | No — different surface. |
| Composio integration | `components/composio/TriggerToggles.tsx`, `composio/providerConfigs.tsx` | Composio integration state | No — different surface. |
| **Channel config** (this PR) | `DiscordConfig.tsx`, `TelegramConfig.tsx` | `channelConnections` slice | **Yes — wired.** |

So this PR's `useOAuthConnectionListener` covers every multi-method OAuth panel that actually exists on the Channels surface. The shared hook is also the right shape for any future `GitHubConfig.tsx` / `GitLabConfig.tsx` channel panels — wiring them in becomes a one-line `useOAuthConnectionListener({ channelId, capabilities, ... })` import.

If the stale-`Connecting` symptom also surfaces in the app-level / skills / Composio OAuth flows, those are separate fixes against different state slices and out of scope for this PR — I'm happy to file follow-up issues if any are observed.

---

### Linear Issue
- Key: N/A
- URL: N/A

### Commit & Branch
- Branch: `fix/2128-oauth-badge-pending-state`
- Commit SHA: `2d93f7c0`

### Validation Run
- [x] `pnpm --filter openhuman-app format:check` — `All matched files use Prettier code style!` on the 6 changed files
- [x] `pnpm typecheck` — clean (`tsc --noEmit`)
- [x] Focused tests: `pnpm --filter openhuman-app exec vitest run --config test/vitest.config.ts src/store/__tests__/channelConnectionsSlice.test.ts src/hooks/__tests__/useOAuthConnectionListener.test.tsx src/components/channels/__tests__/DiscordConfig.test.tsx src/components/channels/__tests__/TelegramConfig.test.tsx` → 4 files, 27 tests pass
- [x] Rust fmt/check (if changed): `N/A: no Rust changes`
- [x] Tauri fmt/check (if changed): `N/A: no Tauri shell changes`

### Validation Blocked
- `command:` `git push` pre-push hook (`app:lint:commands-tokens`)
- `error:` `lint:commands-tokens requires ripgrep` — `rg` not installed on the dev environment
- `impact:` zero — the check greps a directory I did not modify (`src/components/commands/`). Pushed with `--no-verify` per the CLAUDE.md guidance for environment-related hook failures unrelated to the diff. Maintainers can re-run on CI to validate.

### Behavior Changes
- Intended behavior change: OAuth badges on channel panels transition out of `connecting` when the OAuth flow completes *or* fails, and starting a new method cancels the previous method's `connecting` row.
- User-visible effect: the reported bug (multiple methods stuck on `Connecting` simultaneously, Telegram OAuth never clearing) goes away. No new UI elements; only badge state transitions are affected.

### Parity Contract
- Legacy behavior preserved: existing `connected` and `error` transitions are unchanged; `disconnectChannelConnection`, `upsertChannelConnection`, `setChannelConnectionStatus` are all untouched. The Discord `oauth:success` path still produces the same final state (`status: 'connected'`, `capabilities: ['read', 'write']`); the inline effect was just refactored behind the shared hook.
- Guard/fallback/dispatch parity checks: hook only reacts when the event's `toolkit` (success) or `provider` (error) field matches the subscribed channel — siblings on other channels, and mismatched dispatches, are no-ops.

### Duplicate / Superseded PR Handling
- Duplicate PR(s): none found. #2170 cross-references #2128 in passing but its title and body close #2141 (channel selector error-status aggregation, a different surface).
- Canonical PR: this one.
- Resolution: N/A.

## Summary by CodeRabbit

* **New Features**
  * Reusable OAuth connection listener to handle OAuth success/error deep-link flows for Discord and Telegram.
  * New action to clear other pending/connecting auth methods for a channel.

* **Bug Fixes**
  * Prevents multiple auth methods from remaining "connecting"; switching stops in-flight polling and clears sibling pending modes.
  * OAuth errors now record meaningful messages and listeners unsubscribe on unmount.

* **Tests**
  * Added tests covering the OAuth listener and pending-clearing reducer behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tinyhumansai/openhuman/pull/2256?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: app/src/components/channels/DiscordConfig.tsx [hunk 1]
```diff
@@ -153,6 +139,9 @@ const DiscordConfig = ({ definition }: DiscordConfigProps) => {
     (spec: AuthModeSpec) => {
       const key = `discord:${spec.mode}`;
       void runBusy(key, async () => {
+        // Drop any sibling auth mode that's still mid-`connecting` so the
+        // panel doesn't show two methods pinned simultaneously (#2128).
+        dispatch(clearOtherPendingForChannel({ channel: 'discord', exceptAuthMode: spec.mode }));
         dispatch(
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply app/src/components/channels/DiscordConfig.tsx:152]: Fixed in 5936c8d0 — `handleConnect` now unconditionally calls `pollAbort.current?.abort()` and `setLinkToken(null)` at the top, before `clearOtherPendingForChannel`. Safe in the same-mode case because `startLinkPolling` (line 81) immediately spawns a fresh `AbortController`, replacing the just-aborted one. The stale completion path you identified can no longer fire.

File: app/src/components/channels/TelegramConfig.tsx [hunk 2]
```diff
@@ -156,6 +164,9 @@ const TelegramConfig = ({ definition }: TelegramConfigProps) => {
     (spec: AuthModeSpec) => {
       const key = `telegram:${spec.mode}`;
       void runBusy(key, async () => {
+        // Cancel any sibling auth mode still mid-`connecting` so the panel
+        // doesn't pin multiple methods simultaneously (#2128).
+        dispatch(clearOtherPendingForChannel({ channel: 'telegram', exceptAuthMode: spec.mode }));
         dispatch(
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply app/src/components/channels/TelegramConfig.tsx:178]: Fixed in 5936c8d0 — when starting a non-`managed_dm` mode, `handleConnect` calls `stopManagedDmPolling('telegram:managed_dm')` before `clearOtherPendingForChannel`. The same-mode case is already handled inside `startManagedDmPolling` (line 88), which stops any prior controller for the key before creating a new one. Added `stopManagedDmPolling` to the `useCallback` dep array per `react-hooks/exhaustive-deps`.

**Comment by sanil-23** [comment]
## CI status — infra-only failure

After CodeRabbit's approving re-review on `5936c8d0` ([Discord](https://github.com/tinyhumansai/openhuman/pull/2256#discussion_r3269988210) / [Telegram](https://github.com/tinyhumansai/openhuman/pull/2256#discussion_r3269988322)), the two red checks are pure GitHub Actions infrastructure failures, **unrelated to the diff** (which is TypeScript-only — no Rust changes):

- **`test / Rust Core Tests + Quality`** — failed at `cargo metadata` during `Swatinem/rust-cache` post-job cache save with `##[error]No space left on device : '/home/runner/actions-runner/cached/2.334.0/_diag/pages/...'` ([run](https://github.com/tinyhumansai/openhuman/actions/runs/26128679233)). The container was cleaned up before the test step completed.
- **`Rust Core Coverage (cargo-llvm-cov)`** — `cargo llvm-cov` for core ran for 9m24s, then failed in the post-job cache step with `System.IO.IOException: No space left on device : '/home/runner/actions-runner/cached/2.334.0/_diag/Worker_20260519-221722-utc.log'` ([run](https://github.com/tinyhumansai/openhuman/actions/runs/26128679214)). Frontend Coverage and Rust Tauri Coverage on the same workflow both passed.
- **`Coverage Gate (diff-cover ≥ 80%)`** — `skipped` (waiting on Rust Core Coverage's lcov, which never uploaded because of the above).

The first CI run on `2d93f7c0` had all 22 required checks green — same diff up to the small race-condition fix, no Rust touched between runs. I can't `gh run rerun --failed` (no admin on the upstream repo). Could a maintainer re-run those two jobs? That should also un-skip the Coverage Gate.

CodeRabbit's [APPROVED](https://github.com/tinyhumansai/openhuman/pull/2256#pullrequestreview-3338864822) review of the latest commit covers the reviewer-comments gate.

File: app/src/hooks/useOAuthConnectionListener.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * Match key: the event's `toolkit` (success) or `provider` (error) field is
+ * compared case-insensitively to `channel`. Events for other channels are
+ * ignored so multiple panels can mount the hook simultaneously without
+ * stepping on each other.
+ */
+export function useOAuthConnectionListener({
+  channel,
+  authMode,
+  capabilitiesOnSuccess = ['read', 'write'],
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply app/src/hooks/useOAuthConnectionListener.ts]: Fixed in 8a6f6bee — hoisted `DEFAULT_OAUTH_CAPABILITIES = ['read', 'write'] as const` to module scope and switched the default param to reference it. Identity is now stable across renders, so the effect (which depends on `capabilitiesOnSuccess` at line 111) no longer re-subscribes the global `oauth:*` listeners on every parent render. Same class as #2177's daemon-lifecycle fix. Local tests + typecheck + ESLint all green.

**Review by graycyrus — CHANGES_REQUESTED** [review]
Nice work on this one — the shared `useOAuthConnectionListener` hook is a clean extraction, the `clearOtherPendingForChannel` reducer is scoped correctly (only touching `connecting` rows), and all three CodeRabbit findings have been addressed. Tests cover the important paths well.

| File | Area | Summary |
|------|------|---------|
| `useOAuthConnectionListener.ts` | Frontend (new) | Shared hook bridging `oauth:success`/`oauth:error` deep-link events into Redux |
| `channelConnectionsSlice.ts` | Frontend | New `clearOtherPendingForChannel` reducer |
| `DiscordConfig.tsx` | Frontend | Replaced inline OAuth effect with shared hook + sibling cancellation |
| `TelegramConfig.tsx` | Frontend | Added previously-missing OAuth listener + sibling cancellation |
| `useOAuthConnectionListener.test.tsx` | Tests (new) | 8 hook test cases |
| `channelConnectionsSlice.test.ts` | Tests | 4 reducer test cases |

---

**[major] PR-Issue alignment gap — GitHub/GitLab provider coverage**

Issue #2128 acceptance criteria explicitly requires: *"Provider coverage — The fix covers Discord, GitHub, GitLab, and other multi-method OAuth connection panels using the same state path."* This PR only wires `DiscordConfig` and `TelegramConfig`. If GitHub/GitLab channel configs have OAuth auth modes going through the same `channelConnections` slice, they need `useOAuthConnectionListener` mounted too — otherwise the issue's acceptance criteria aren't met and those panels remain vulnerable to the same stale-badge bug.

The shared hook makes this straightforward to add. If you've verified those channels don't have the OAuth badge issue (e.g. they use a different state path or don't have multi-method OAuth), a note in the PR description explaining why they're excluded would close this out.

---

Everything else looks solid. The hook's channel-matching logic, the module-level `DEFAULT_OAUTH_CAPABILITIES` constant, the poll-abort ordering in both configs, and the `disconnected` (not `error`) transition for cancelled siblings are all good choices.

**Review by M3gA-Mind — APPROVED** [review]

**Comment by sanil-23** [comment]
@graycyrus thanks for the read. You called it correctly — and the answer is the second option you offered: GitHub / GitLab don't have channel-config panels in this codebase today, so there's nothing to wire the new hook into for those providers.

I've added a **Provider coverage** section to the PR description with the full evidence. Short version:

- `find app/src -name "*Config.tsx"` returns three channel configs total: `DiscordConfig.tsx`, `TelegramConfig.tsx`, `WebChannelConfig.tsx` (the last is not OAuth-driven). No `GitHubConfig.tsx` / `GitLabConfig.tsx` anywhere.
- GitHub OAuth does exist elsewhere — `BootCheckGate.tsx` (app-level sign-in), `InstallSkillDialog.tsx` (skill OAuth), `composio/TriggerToggles.tsx` (Composio integration) — but each lives on a different Redux/state slice (`deepLinkAuth`, skills-domain, Composio-integration) and doesn't share the `channelConnections` code path this PR's hook is bound to.
- The shared `useOAuthConnectionListener` is the right shape for any future `GitHubConfig.tsx` / `GitLabConfig.tsx` if they get added — wiring becomes a single import.

If the stale-`Connecting` symptom turns out to surface in the app-level sign-in or skill / Composio OAuth flows (with their own stuck-state mechanics on those other slices), I'd treat each as a separate fix against the relevant slice rather than expanding this PR.

Let me know if the body update lands the concern or if you'd rather see the missing-panel detection asserted in a regression test as well.

**Comment by senamakel** [comment]
huge thanks @sanil-23, this one's a gem 🙌 centralising the oauth badge transitions behind that new listener hook is such a clean fix, no more stale "connecting" badges hanging around ✨ really appreciate you back here again 💚
